### PR TITLE
Add Mellanox SRIOV hw-offload support

### PR DESCRIPTION
**Pull request by pperiyasamy**
This PR provides an option to enable OVS hardware offload with ovs-cni
while using OVS data-plane in Mellanox ConnectX-4 Lx onwards NIC
hardware (Mellanox Embedded Switch or eSwitch).  It helps to accelerate
the networking throughput for the applications running inside the pod
container.

The 1st commit adds smart VF representor into ovs bridge. This way
network configurations like vlan, ipam etc. can be configured for the
smart VF attached inside pod container.

The unit tests would get added after a review of this commit.

Signed-off-by: Periyasamy Palanisamy <periyasamy.palanisamy@est.tech>

### CODE REVIEW

**Comment by ovirt-infra** [comment]
Hello contributor, thanks for submitting a PR for this project! 

I am the bot who triggers "standard-CI" builds for this project.
As a security measure, I will not run automated tests on PRs that are not from white-listed contributors.

In order to allow automated tests to run, please ask one of the project maintainers to review the code and then do one of the following:

1. Type `ci test please` on this PR to trigger automated tests for it.
2. Type `ci add to whitelist` on this PR to trigger automated tests for it and also add you to the contributor white-list so that your future PRs will be tested automatically. ( keep in mind this list might be overwritten if the job XML is refreshed, for permanent whitelisting, please follow #3 option )
3. If you are planning to contribute to more than one project, maybe it's better to ask them to add you to the project organization, so you'll be able to run tests for all the organization's projects.

**Comment by pperiyasamy** [comment]
#https://github.com/ovn-org/ovn-kubernetes/pull/665

**Comment by phoracek** [comment]
Could you please add the doc to https://github.com/kubevirt/ovs-cni/blob/master/docs/cni-plugin.md?

**Comment by pperiyasamy** [comment]
> Could you please add the doc to https://github.com/kubevirt/ovs-cni/blob/master/docs/cni-plugin.md?

@phoracek Sure, will update the doc. Meanwhile you could check ovs offload support at https://docs.google.com/document/d/11uBezjSZg5b2g6VNs2mugGrgK5ZX9c-Xirj6XL_acjM/edit

**Comment by dankenigsberg** [comment]
@ pperiyasamy  thanks; would you also update the first conversation message with a human-facing description of what you aim to do, and why it is relevant to ovs-cni?

File: docs/cni-plugin.md [hunk 1]
```diff
@@ -92,3 +92,87 @@ This also needs `host-local` ipam plugin to be present in one of the `PATH` dire
...
+metadata:
+  name: ovs-trunk-net
+  annotations:
+    k8s.v1.cni.cncf.io/resourceName: intel.com/mellanox_snic0
+spec:
+  config: '{
+      "cniVersion": "0.3.1",
+      "type": "ovs",
+      "bridge": "br-snic0",
+      "trunk": "1050-1059"
```
Comment (JanScheurich) [inline docs/cni-plugin.md]: The trunk attribute syntax is not aligned with the merged version of the earlier pull request, which uses structured json to specify a list of ranges of VLAN IDs.

File: docs/cni-plugin.md [hunk 2]
```diff
@@ -92,3 +92,87 @@ This also needs `host-local` ipam plugin to be present in one of the `PATH` dire
...
+    }'
+```
+
+Now deploy a pod with the following config to attach VF into container and its representor
```
Comment (dankenigsberg) [inline docs/cni-plugin.md]: Who is expected to create `br-snic0`? How? How is a correct VF attached to that bridge?
Can you provide an example? Is there a way to do it in a k8s-friendly way, e.g. via https://github.com/nmstate/kubernetes-nmstate ?

File: docs/cni-plugin.md [hunk 3]
```diff
@@ -92,3 +92,87 @@ This also needs `host-local` ipam plugin to be present in one of the `PATH` dire
...
+
+### Mellanox SR-IOV Configuation
+
+To enable OVS hardware offloading, create VFs from a Mellanox PF, configure VFs in `switchdev` mode and
+set `hw-offload=true` for Open vSwitch.
+For more information [refer Mellanox doc] (https://www.mellanox.com/related-docs/prod_software/ASAP2_Hardware_Offloading_for_vSwitches_User_Manual_v4.4.pdf)
+
+### Device Plugin configuation
+
+The device plugin would create the resouce pools based on the configurations given in the `/etc/pcidp/config.json`.
```
Comment (dankenigsberg) [inline docs/cni-plugin.md]: who is expected to create this file properly? would  https://github.com/openshift/sriov-network-operator create something acceptable?

File: docs/cni-plugin.md (hunk 2 above)
Comment (JanScheurich) [reply docs/cni-plugin.md]: @dankenigsberg: There is no fundamental difference to the current veth-pair based network attachments. An ovs-cni NAD specifies an OVS bridge representing a "physical network" hosting the VLANs. 

In a bare metal deployment this bridge would be connected to the physical network through an OVS bond for redundancy. With Mellanox SmartNICs, the OVS bridge is connected to a Linux bond of the two SmartNIC PFs, as that is the way Mellanox supports OVS HW offload with LAG through the TC flower API.

The ovs-cni assumes that these OVS bridges have been set up and that the VLANs have been provisioned on the physical networks connecting the K8s workers prior to the creation of the NADs.

Automation of such network configuration is indeed highly wanted and I am very grateful for the pointer to the kubernetes-nmstate project. We will have a look there.

File: docs/cni-plugin.md (hunk 3 above)
Comment (JanScheurich) [reply docs/cni-plugin.md]: There is a close connection between the configuration of (SRIOV) VF device pools and the creation of NADs referring to these as resources (for automatic injection of resources into the pod spec through the network resources injector (https://github.com/intel/network-resources-injector). For a given Mellanox SmartNICs with LAG, all VFs will be typically form one single pool as the they are fully "programmable" as OVS interfaces at pod creation. 

So a typical setup would be CX-5 SmartNIC with two bonded PFs  associated with one OVS bridge and one VF pool. On top of that there could be any number of ovs-cni NADs for individual VLAN or VLAN trunk ports on that bridge.

We have been looking at the openshift sriov-network-operator before, but found it not fulfilling our needs fully. Also there were some dependencies on the openshift environment that made it hard to get it to work on vanilla K8s.

I think we should look into the provisioning of networks and associated network device pools coherently in one effort.

File: pkg/plugin/plugin.go [hunk 4]
```diff
@@ -114,6 +116,125 @@ func generateRandomMac() net.HardwareAddr {
 	return net.HardwareAddr(append(prefix, suffix...))
 }
 
+func setupSriovInterface(contNetns ns.NetNS, ifName string, mtu int, pciAddrs string) (*current.Interface, *current.Interface, error) {
```
Comment (phoracek) [inline pkg/plugin/plugin.go]: s/pciAddrs/deviceID ?

File: pkg/plugin/plugin.go [hunk 5]
```diff
@@ -114,6 +116,125 @@ func generateRandomMac() net.HardwareAddr {
...
+		return nil, nil, err
+	}
+
+	hostIface.Name = rep
+
+	link, err := netlink.LinkByName(hostIface.Name)
+	if err != nil {
+		return nil, nil, err
+	}
+	hostIface.Mac = link.Attrs().HardwareAddr.String()
```
Comment (phoracek) [inline pkg/plugin/plugin.go]: Is it possible to change this MAC address? Would it make sense?

File: pkg/plugin/plugin.go [hunk 6]
```diff
@@ -114,6 +116,125 @@ func generateRandomMac() net.HardwareAddr {
...
+		return nil, nil, err
+	}
+
+	// Make sure we have 1 netdevice per pci address
+	if len(vfNetdevices) != 1 {
+		return nil, nil, fmt.Errorf("failed to get one netdevice interface per %s", pciAddrs)
+	}
+	vfNetdevice := vfNetdevices[0]
+
+	// 2. get Uplink netdevice
```
Comment (phoracek) [inline pkg/plugin/plugin.go]: Could you please explain in a comment what is uplink representor?

File: pkg/plugin/plugin.go [hunk 7]
```diff
@@ -114,6 +116,125 @@ func generateRandomMac() net.HardwareAddr {
...
+		return nil, nil, err
+	}
+
+	// 3. get VF index from PCI
+	vfIndex, err := sriovnet.GetVfIndexByPciAddress(pciAddrs)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	// 4. lookup representor
```
Comment (phoracek) [inline pkg/plugin/plugin.go]: could you please explain in a comment what is vf representor?

File: docs/cni-plugin.md (hunk 2 above)
Comment (dankenigsberg) [reply docs/cni-plugin.md]: Thanks. I'd appreciate if something of this great explanation finds it place into README.md.

File: docs/cni-plugin.md (hunk 1 above)
Comment (pperiyasamy) [reply docs/cni-plugin.md]: done

File: docs/cni-plugin.md (hunk 2 above)
Comment (pperiyasamy) [reply docs/cni-plugin.md]: Sure, i've now moved it into a new document.

File: pkg/plugin/plugin.go [hunk 8]
```diff
@@ -114,6 +116,125 @@ func generateRandomMac() net.HardwareAddr {
 	return net.HardwareAddr(append(prefix, suffix...))
 }
 
+func setupSriovInterface(contNetns ns.NetNS, ifName string, mtu int, pciAddrs string) (*current.Interface, *current.Interface, error) {
+	hostIface := &current.Interface{}
+	contIface := &current.Interface{}
+
+	// 1. get VF netdevice from PCI
```
[the remark this answers is not in the document]
Comment (pperiyasamy) [reply pkg/plugin/plugin.go]: done

File: pkg/plugin/plugin.go [hunk 9]
```diff
@@ -309,7 +430,15 @@ func CmdAdd(args *skel.CmdArgs) error {
 	}
 	defer contNetns.Close()
 
-	hostIface, contIface, err := setupVeth(contNetns, args.IfName, mac, netconf.MTU)
+	var hostIface, contIface *current.Interface
+	if netconf.DeviceID != "" {
+		// SR-IOV Case
+		hostIface, contIface, err = setupSriovInterface(contNetns, args.IfName, netconf.MTU, netconf.DeviceID)
+	} else {
+		//General Case
```
[the remark this answers is not in the document]
Comment (pperiyasamy) [reply pkg/plugin/plugin.go]: done

File: docs/cni-plugin.md [hunk 10]
```diff
@@ -92,3 +92,87 @@ This also needs `host-local` ipam plugin to be present in one of the `PATH` dire
...
+
+```yaml
+apiVersion: "k8s.cni.cncf.io/v1"
+kind: NetworkAttachmentDefinition
+metadata:
+  name: ovs-trunk-net
+  annotations:
+    k8s.v1.cni.cncf.io/resourceName: intel.com/mellanox_snic0
+spec:
+  config: '{
```
[the remark this answers is not in the document]
Comment (pperiyasamy) [reply docs/cni-plugin.md]: @phoracek The deviceID is not required when device plugin and multus are used. 
The multus makes sure that assigned device's pci address is passed as deviceID
parameter into ovs-cni plugin.

File: docs/cni-plugin.md [hunk 11]
```diff
@@ -92,3 +92,87 @@ This also needs `host-local` ipam plugin to be present in one of the `PATH` dire
...
+### Prerequisites
+
+[sriov-network-device-plugin](https://github.com/intel/sriov-network-device-plugin)
+[multus-cni](https://github.com/intel/multus-cni)
+[network-resources-injector](https://github.com/intel/network-resources-injector)
+
+### Mellanox SR-IOV Configuation
+
+To enable OVS hardware offloading, create VFs from a Mellanox PF, configure VFs in `switchdev` mode and
+set `hw-offload=true` for Open vSwitch.
```
[the remark this answers is not in the document]
Comment (pperiyasamy) [reply docs/cni-plugin.md]: done

File: pkg/plugin/plugin.go (hunk 4 above)
Comment (pperiyasamy) [reply pkg/plugin/plugin.go]: done

File: pkg/plugin/plugin.go (hunk 5 above)
Comment (pperiyasamy) [reply pkg/plugin/plugin.go]: This is just for completion purpose not intended to change the mac address.
It just assigns the same existing VF mac address into hostIface.Mac.

File: pkg/plugin/plugin.go (hunk 6 above)
Comment (pperiyasamy) [reply pkg/plugin/plugin.go]: done

File: pkg/plugin/plugin.go (hunk 7 above)
Comment (pperiyasamy) [reply pkg/plugin/plugin.go]: done, Hope that's fine for you now :)

File: docs/cni-plugin.md (hunk 10 above)
[the remark this answers is not in the document]
Comment (phoracek) [reply docs/cni-plugin.md]: Neat!

File: pkg/plugin/plugin.go [hunk 12]
```diff
@@ -560,17 +569,58 @@ func CmdDel(args *skel.CmdArgs) error {
...
 
-	// Delete can be called multiple times, so don't return an error if the
-	// device is already removed.
-	err = ns.WithNetNSPath(args.Netns, func(ns.NetNS) error {
-		err = ip.DelLinkByName(args.IfName)
-		if err != nil && err == ip.ErrLinkNotFound {
-			return nil
-		}
+	if netconf.DeviceID != "" {
+		//  SR-IOV Case
```
Comment (phoracek) [inline pkg/plugin/plugin.go]: Can we call it multiple times (like with the general case)?

File: pkg/utils/utils.go [hunk 13]
```diff
@@ -0,0 +1,123 @@
...
+	err := ioutil.WriteFile(path, conf, 0600)
+	if err != nil {
+		return fmt.Errorf("failed to write container data in the path(%q): %v", path, err)
+	}
+
+	return err
+}
+
+// ReadScratchConf takes in container ID, Pod interface name and data dir as string and returns a pointer to Conf
+func ReadScratchConf(cRefPath string) ([]byte, error) {
```
Comment (phoracek) [inline pkg/utils/utils.go]: can we make this into a private function?

File: pkg/utils/utils.go [hunk 14]
```diff
@@ -0,0 +1,123 @@
+// Licensed under the Apache License, Version 2.0 (the "License");
```
Comment (phoracek) [inline pkg/utils/utils.go]: I really hate bloating utils.go modules. Could you please split it into one module for the VF handling and one for generic cache? Or at least rename it to something that would suggest this utils module is VF specific.

File: pkg/plugin/plugin.go (hunk 12 above)
Comment (pperiyasamy) [reply pkg/plugin/plugin.go]: sure, added a check to handle `ErrLinkNotFound` error.

File: pkg/utils/utils.go (hunk 13 above)
Comment (pperiyasamy) [reply pkg/utils/utils.go]: done.

File: pkg/utils/utils.go (hunk 14 above)
Comment (pperiyasamy) [reply pkg/utils/utils.go]: Yes, makes sense. moved into sriov package having sriov and cache go modules.

**Pull request merged**